hbase-server/src/main/java/org/apache/hadoop/hbase/master/zksyncer/ClientZKSyncer.java

```diff
@@ -34,30 +33,69 @@
 import org.apache.yetus.audience.InterfaceAudience;
 import org.apache.zookeeper.CreateMode;
 import org.apache.zookeeper.KeeperException;
-
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
  * Tracks the target znode(s) on server ZK cluster and synchronize them to client ZK cluster if
  * changed
  * <p/>
- * The target znode(s) is given through {@link #getNodesToWatch()} method
+ * The target znode(s) is given through {@link #getPathsToWatch()} method
  */
 @InterfaceAudience.Private
 public abstract class ClientZKSyncer extends ZKListener {
   private static final Logger LOG = LoggerFactory.getLogger(ClientZKSyncer.class);
   private final Server server;
   private final ZKWatcher clientZkWatcher;
+
+  private static final class ZKData {
+
+    byte[] data;
+
+    boolean delete = false;
+
+    synchronized void set(byte[] data) {
+      this.data = data;
+      notifyAll();
+    }
+
+    synchronized byte[] get() throws InterruptedException {
+      while (!delete && data == null) {
+        wait();
+      }
+      byte[] d = data;
+      data = null;
+      return d;
+    }
+
+    synchronized void delete() {
+      this.delete = true;
+      notifyAll();
+    }
+
+    synchronized boolean isDeleted() {
+      return delete;
+    }
+  }
+
   // We use queues and daemon threads to synchronize the data to client ZK cluster
```

Comment:
The newly introduced `ZKData` structure is more efficient, shall we update the comments here and add some javadoc for the `ZKData` class?